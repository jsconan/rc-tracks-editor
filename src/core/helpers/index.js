/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

export { default as adjust } from './adjust.js';
export { default as alternate } from './alternate.js';
export { default as assign } from './assign.js';
export { default as defined } from './defined.js';
export { default as eachWord } from './eachWord.js';
export { default as getOwnPropertyDescriptors } from './getOwnPropertyDescriptors.js';
export { default as hasAPI } from './hasAPI.js';
export { default as mixin } from './mixin.js';
export { default as pick } from './pick.js';
export { default as uid } from './uid.js';
export { default as validateAPI } from './validateAPI.js';
export { default as validateCallback } from './validateCallback.js';
export { default as wait } from './wait.js';
export * from './angles.js';
export * from './compare.js';
export * from './svg.js';
export * from './transform.js';
