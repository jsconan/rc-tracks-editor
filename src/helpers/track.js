/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { RIGHT_ANGLE } from './maths';

/**
 * Computes the overall width of a track section.
 * @param {number} laneWidth - The width of a track lane.
 * @param {number} barrierWidth - The width of the barriers.
 * @returns {number}
 */
export const getTrackSectionWidth = (laneWidth, barrierWidth) => laneWidth + barrierWidth * 2;

/**
 * Computes the overall length of a track section.
 * @param {number} laneWidth - The width of a track lane.
 * @param {number} barrierWidth - The width of the barriers.
 * @returns {number}
 */
export const getTrackSectionLength = (laneWidth, barrierWidth) =>
    getTrackSectionWidth(laneWidth, barrierWidth) + laneWidth / 4;

/**
 * Computes the width of the track lane from the given track section length and width.
 * @param {number} width - The width of a track section.
 * @param {number} barrierWidth - The width of the barriers.
 * @returns {number}
 */
export const getTrackLaneWidth = (width, barrierWidth) => width - barrierWidth * 2;

/**
 * Computes the angle of a curve with respect to the ratio.
 * @param {number} ratio - The ratio of the curve.
 * @returns {number}
 */
export const getCurveAngle = ratio => RIGHT_ANGLE / (ratio < 1 ? 1 / ratio : ratio);

/**
 * Computes the inner radius of a curve given the ratio.
 * @param {number} length - The length of a track section.
 * @param {number} width - The width of a track section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getCurveInnerRadius = (length, width, ratio = 1) => length * (ratio - 1) + (length - width) / 2;

/**
 * Computes the outer radius of a curve given the ratio.
 * @param {number} length - The length of a track section.
 * @param {number} width - The width of a track section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getCurveOuterRadius = (length, width, ratio = 1) => width + getCurveInnerRadius(length, width, ratio);

/**
 * Computes the length of the outer side of an enlarged curved track.
 * @param {number} length - The length of a track section.
 * @param {number} width - The width of a track section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getEnlargedCurveSide = (length, width, ratio = 1) => (length * ratio) / 2;

/**
 * Computes the inner radius of a curve given the ratio.
 * @param {number} length - The length of a track section.
 * @param {number} width - The width of a track section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getEnlargedCurveInnerRadius = (length, width, ratio = 1) => getCurveInnerRadius(length, width, ratio);

/**
 * Computes the outer radius of an enlarged curved track.
 * @param {number} length - The length of a track section.
 * @param {number} width - The width of a track section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getEnlargedCurveOuterRadius = (length, width, ratio = 1) =>
    getCurveOuterRadius(length, width, ratio) - getEnlargedCurveSide(length, width, ratio);

/**
 * Computes the number of barrier chunks for a straight section given the ratio.
 * @param {number} barrierChunks - The number of barrier chunks per section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getStraightBarrierChunks = (barrierChunks, ratio = 1) => barrierChunks * ratio;

/**
 * Computes the number of barrier chunks for an inner curved section given the ratio.
 * @param {number} barrierChunks - The number of barrier chunks per section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getCurveInnerBarrierChunks = (barrierChunks, ratio = 1) =>
    ratio < 1 ? 1 : barrierChunks / (ratio < 2 ? 2 : 1);

/**
 * Computes the number of barrier chunks for an outer curved section given the ratio.
 * @param {number} barrierChunks - The number of barrier chunks per section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getCurveOuterBarrierChunks = (barrierChunks, ratio = 1) => barrierChunks / (ratio < 1 ? 2 : 1);

/**
 * Computes the number of barrier chunks for the straight sides of large curve track.
 * @param {number} barrierChunks - The number of barrier chunks per section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getEnlargedCurveSideBarrierChunks = (barrierChunks, ratio = 1) =>
    getStraightBarrierChunks(barrierChunks, ratio) / 2;

/**
 * Computes the number of barrier chunks for the inner curve of large curve track.
 * @param {number} barrierChunks - The number of barrier chunks per section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getEnlargedCurveInnerBarrierChunks = (barrierChunks, ratio = 1) =>
    ratio === 1 ? barrierChunks / 2 : barrierChunks * ratio;

/**
 * Computes the number of barrier chunks for the outer curve of large curve track.
 * @param {number} barrierChunks - The number of barrier chunks per section.
 * @param {number} [ratio] - The size factor.
 * @returns {number}
 */
export const getEnlargedCurveOuterBarrierChunks = (barrierChunks, ratio = 1) => (barrierChunks / 2) * ratio;
