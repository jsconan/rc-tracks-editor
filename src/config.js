/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { getTrackSectionWidth, getTrackSectionLength } from './helpers/track.js';

/**
 * The dimensions and constraints of a track element
 * @type {object}
 */
const config = {
    /**
     * The width of the track lane (the distance between the barriers).
     *
     * This value can be modified.
     *
     * @type {number}
     */
    trackLaneWidth: 120,

    /**
     * The width of the barriers.
     *
     * This value can be modified.
     *
     * @type {number}
     */
    barrierWidth: 6,

    /**
     * The number of barrier chunks per section.
     *
     * This value can be modified.
     *
     * @type {number}
     */
    barrierChunks: 4,

    /**
     * The overall length of a track section (size of a tile in the track).
     *
     * DO NOT MODIFY! This value will be generated from the compute() method.
     *
     * @type {number}
     */
    trackSectionLength: 0,

    /**
     * The overall width of a track section (size of a tile in the track).
     *
     * DO NOT MODIFY! This value will be generated from the compute() method.
     *
     * @type {number}
     */
    trackSectionWidth: 0,

    /**
     * Generate values from the base configuration.
     * The generated values are: `trackSectionLength`, `trackSectionWidth`, `barrierLength`.
     * @returns {object} - Returns the configuration.
     */
    compute() {
        this.trackSectionLength = getTrackSectionLength(this.trackLaneWidth, this.barrierWidth);
        this.trackSectionWidth = getTrackSectionWidth(this.trackLaneWidth, this.barrierWidth);
        return this;
    }
};

export default config.compute();
