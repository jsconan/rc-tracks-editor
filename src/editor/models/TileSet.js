/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { getTypes, TILE_DIRECTION_RIGHT } from '../../tile/helpers';
import { TileCounter, TileList } from '../../tile/models';
import { TileSpecifications } from '../../tile/config';
import { TileCounterStore, TileModelsStore } from '../../tile/stores';

/**
 * Represents a set of tile models that can be consumed.
 */
export class TileSet extends TileCounter {
    /**
     * Represents a set of tile models that can be consumed.
     * @param {TileSpecifications} specs - The specifications for the track tiles.
     * @param {TileModel[]} source - A list of tiles to initialize the set.
     * @throws {TypeError} - If the given specifications object is not valid.
     * @throws {TypeError} - If one of the tiles in the source is not a TileModel.
     */
    constructor(specs, source = null) {
        super(source);

        this.counter = new TileCounter();
        this.setSpecs(specs);

        if (!source) {
            this.loadAllModels();
        }
        source = void 0;

        this.modelsStore = new TileModelsStore(this);
        this.counterStore = new TileCounterStore(this);

        this.on('removetile', (tile, amount) => this.counter.add(tile, amount));
        this.on('addtile', (tile, amount) => this.counter.remove(tile, amount));
    }

    /**
     * Sets the specifications for the track tiles.
     * @param {TileSpecifications} specs - The specifications for the track tiles.
     * @returns {TileSet} - Chains the instance.
     * @throws {TypeError} - If the given specifications object is not valid.
     * @fires specs
     */
    setSpecs(specs) {
        TileSpecifications.validateInstance(specs);

        this.specs = specs;
        this.counter.setSpecs(specs);

        return super.setSpecs(specs);
    }

    /**
     * Restores the tiles removed from the set.
     * @returns {TileSet} - Chains the instance.
     * @fires set
     * @fires addmodel
     * @fires addtile
     * @fires reset
     */
    reset() {
        this.modelsStore.unbind();
        this.counterStore.unbind();

        for (const { model, count } of this.counter.getCounterList()) {
            this.add(model, count);
        }
        this.counter.clear();

        this.modelsStore.bind(this);
        this.counterStore.bind(this);

        this.emit('reset');

        return this;
    }

    /**
     * Consumes tiles from the set.
     * @param {TileModel[]} tiles - A list of tiles that is using the set.
     * @returns {TileSet} - Chains the instance.
     * @fires set
     * @fires removemodel
     * @fires delete
     * @fires removetile
     * @fires consume
     */
    consume(tiles) {
        this.modelsStore.unbind();
        this.counterStore.unbind();

        for (const tile of tiles) {
            this.remove(tile);
        }

        this.modelsStore.bind(this);
        this.counterStore.bind(this);

        this.emit('consume');

        return this;
    }

    /**
     * Loads counters from the list of all available tile models.
     * The counters are cleared before, including the related tile models.
     * @param {number} count - The initial count given to each counter.
     * @returns {TileSet} - Chains the instance.
     * @fires load
     */
    loadAllModels(count = Number.POSITIVE_INFINITY) {
        /**
         * @param {TileCounter} counter
         * @yields {Array}
         * @generator
         * @private
         */
        function* loadAll(counter) {
            for (const type of getTypes()) {
                let ratio = 1;
                let maxRatio = 1;
                do {
                    const model = TileList.createTile(counter.specs, type, TILE_DIRECTION_RIGHT, ratio);

                    yield [model, count];

                    maxRatio = model.maxRatio;
                } while (ratio++ < maxRatio);
            }
        }

        return this.load(loadAll(this));
    }

    /**
     * Loads counters from a list of tiles.
     * The counters are cleared before, including the related tile models.
     * @param {*} iterator - An iterable object that can be used to set the counters.
     * @returns {TileSet} - Chains the instance.
     * @fires load
     */
    load(iterator) {
        if (!iterator || !iterator[Symbol.iterator]) {
            return this;
        }

        if (this.counter) {
            this.counter.clear();
        }
        return super.load(iterator);
    }

    /**
     * Removes all counters, including the related tile models.
     * @returns {TileSet} - Chains the instance.
     * @fires clear
     */
    clear() {
        this.counter.clear();

        return super.clear();
    }
}

/**
 * @typedef {import('../../tile/models').TileModel} TileModel
 */

/**
 * Notifies the tile specification have changed.
 * @event specs
 * @param {TileSpecifications} specs - The specifications for the tiles.
 */

/**
 * Notifies a tile model has been added.
 * @event addmodel
 * @param {string} key - The key of the tile model that was added.
 * @param {TileModel} model - The tile model that was added.
 */

/**
 * Notifies a tile model has been removed.
 * @event removemodel
 * @param {string} key - The key of the tile model that was removed.
 * @param {TileModel} model - The tile model that was removed.
 */

/**
 * Notifies a tile has been added.
 * @event addtile
 * @param {TileModel} tile - The tile that was added.
 * @param {number} amount - The number of similar tiles that was added.
 */

/**
 * Notifies a tile has been removed.
 * @event removetile
 * @param {TileModel} tile - The tile that was removed.
 * @param {number} amount - The number of similar tiles that was removed.
 */

/**
 * Notifies a value has been set to a counter.
 * @event set
 * @param {string} key - The key of the counter that was set.
 * @param {number} newValue - The new value.
 * @param {number} oldValue - The previous value.
 */

/**
 * Notifies a counter has been removed.
 * @event delete
 * @param {string} key - The key of the counter that was removed.
 * @param {number} value - The last value of the removed counter.
 */

/**
 * Notifies the removed tiles have been restored to the set.
 * @event reset
 */

/**
 * Notifies a list of tiles have been consumed from the set.
 * @event consume
 */

/**
 * Notifies all counters has been removed.
 * @event clear
 */

/**
 * Notifies the counters has been loaded.
 * @event load
 */
