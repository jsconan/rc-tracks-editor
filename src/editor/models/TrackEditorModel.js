/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import eventEmitterMixin from '../../core/mixins/eventEmitter.js';
import { TileSet } from './TileSet.js';
import { TileSpecifications } from '../../tile/config';
import { TrackModel } from '../../track/models';

/**
 * Represent a track editor
 */
export class TrackEditorModel {
    /**
     * Creates a track editor model.
     * @param {TileSpecifications} specs - The specifications for the track tiles.
     * @param {number} angle - A rotation angle for the track.
     */
    constructor(specs, angle = 0) {
        eventEmitterMixin(this);

        this.tiles = new TileSet(specs);
        this.track = new TrackModel(specs);
        this.specs = specs;

        this.track.setBuilderOptions({
            startAngle: angle,
            hPadding: specs.padding,
            vPadding: specs.padding
        });

        this.track.on('add', (...tiles) => tiles.forEach(tile => this.tiles.remove(tile)));
        this.track.on('remove', (...tiles) => tiles.forEach(tile => this.tiles.add(tile)));
        this.track.on('clear', () => this.tiles.reset());
        this.track.on('load', () => {
            this.tiles.reset();
            this.tiles.consume(this.track);
        });
    }

    /**
     * Sets the specifications for the track tiles.
     * @param {TileSpecifications} specs - The specifications for the track tiles.
     * @returns {TrackEditorModel} - Chains the instance.
     * @throws {TypeError} - If the given specifications object is not valid.
     * @fires specs
     */
    setSpecs(specs) {
        TileSpecifications.validateInstance(specs);

        this.specs = specs;
        this.tiles.setSpecs(specs);
        this.track.setSpecs(specs);

        this.emit('specs', specs);

        return this;
    }

    /**
     * Sets the rotation angle for the track.
     * @param {number} angle - A rotation angle for the track.
     * @returns {TrackEditorModel} - Chains the instance.
     */
    setAngle(angle) {
        this.track.setBuilderOption('startAngle', angle);
        return this;
    }

    /**
     * Limits the set of available tiles for building a track.
     * @param {*} tiles - An iterable object that can be used to set the tiles.
     * @returns {TrackEditorModel} - Chains the instance.
     * @fires limit
     */
    limit(tiles = null) {
        if (tiles) {
            this.tiles.load(tiles);
        } else {
            this.tiles.loadAllModels();
        }

        if (this.track.length) {
            this.tiles.consume(this.track);
        }

        this.emit('limit');

        return this;
    }

    /**
     * Loads a track.
     * @param {tileExport[]} data - An iterable object containing a representation of the track.
     * @returns {TrackEditorModel} - Chains the instance.
     * @fires load
     */
    load(track) {
        this.track.import(track);

        this.emit('load');

        return this;
    }

    /**
     * Saves the track.
     * @returns {tileExport[]} - An object representation of the track.
     * @fires save
     */
    save() {
        const track = this.track.export();

        this.emit('save', track);

        return track;
    }

    /**
     * Removes all tiles from the track, and reset the set of tiles.
     * @returns {TrackEditorModel} - Chains the instance.
     * @fires clear
     */
    clear() {
        this.track.clear();

        this.emit('clear');

        return this;
    }
}

/**
 * @typedef {import('../../tile/models/TileModel.js').tileExport} tileExport
 */

/**
 * Notifies the tile specification have changed.
 * @event specs
 * @param {TileSpecifications} specs - The specifications for the tiles.
 */

/**
 * Notifies the set of available tiles has been limited
 * @event limit
 */

/**
 * Notifies a track has been loaded.
 * @event load
 */

/**
 * Notifies the track has been saved.
 * @event save
 * @param {tileExport[]} track - An iterable object containing a representation of the track.
 */

/**
 * Notifies the editor has been cleared.
 * @event clear
 */
