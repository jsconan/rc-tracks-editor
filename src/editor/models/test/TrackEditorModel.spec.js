/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { CURVED_TILE_TYPE, STRAIGHT_TILE_TYPE, TILE_DIRECTION_LEFT, TILE_DIRECTION_RIGHT } from '../../../tile/helpers';
import { CurvedTileEnlargedModel, CurvedTileModel, StraightTileModel } from '../../../tile/models';
import { TileSpecifications } from '../../../tile/config';
import { TileSet } from '../TileSet';
import { TrackEditorModel } from '../TrackEditorModel.js';
import { TrackModel } from '../../../track/models';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;
const maxRatio = 4;
const specs = new TileSpecifications({ laneWidth, barrierWidth, barrierChunks, maxRatio });
const models = [
    new StraightTileModel(specs),
    new CurvedTileModel(specs, CurvedTileModel.DIRECTION_RIGHT, 1),
    new CurvedTileModel(specs, CurvedTileModel.DIRECTION_RIGHT, 2),
    new CurvedTileModel(specs, CurvedTileModel.DIRECTION_RIGHT, 3),
    new CurvedTileModel(specs, CurvedTileModel.DIRECTION_RIGHT, 4),
    new CurvedTileEnlargedModel(specs)
];
const unlimitedTiles = [
    [models[0].modelId, Number.POSITIVE_INFINITY],
    [models[1].modelId, Number.POSITIVE_INFINITY],
    [models[2].modelId, Number.POSITIVE_INFINITY],
    [models[3].modelId, Number.POSITIVE_INFINITY],
    [models[4].modelId, Number.POSITIVE_INFINITY],
    [models[5].modelId, Number.POSITIVE_INFINITY]
];
const limitedTiles = [
    [new StraightTileModel(specs), 2],
    [new CurvedTileEnlargedModel(specs), 2],
    [new CurvedTileModel(specs), 2]
];
const track = [
    {
        type: CURVED_TILE_TYPE,
        direction: TILE_DIRECTION_LEFT,
        ratio: 1
    },
    {
        type: STRAIGHT_TILE_TYPE,
        direction: TILE_DIRECTION_RIGHT,
        ratio: 1
    },
    {
        type: CURVED_TILE_TYPE,
        direction: TILE_DIRECTION_RIGHT,
        ratio: 1
    }
];

describe('TrackEditorModel', () => {
    it('is a class', () => {
        expect(TrackEditorModel).toEqual(expect.any(Function));
    });

    it('needs a specifications object', () => {
        const editor = new TrackEditorModel(specs);

        expect(editor.specs).toBe(specs);

        expect(() => new TrackEditorModel()).toThrow(
            'The specifications object must be an instance of TileSpecifications!'
        );
        expect(() => new TrackEditorModel({})).toThrow(
            'The specifications object must be an instance of TileSpecifications!'
        );
    });

    it('has a tiles property', () => {
        const editor = new TrackEditorModel(specs);

        expect(editor.tiles).toBeInstanceOf(TileSet);
    });

    it('has a track property', () => {
        const editor = new TrackEditorModel(specs);

        expect(editor.track).toBeInstanceOf(TrackModel);
    });

    it('is created with a default set of tiles', () => {
        const editor = new TrackEditorModel(specs);

        expect([...editor.tiles]).toEqual(unlimitedTiles);
    });

    it('is created with an empty track', () => {
        const editor = new TrackEditorModel(specs);

        expect([...editor.track]).toStrictEqual([]);
    });

    it('can set the specifications of the tile models', () => {
        const editor = new TrackEditorModel(specs);
        const newSpecs = new TileSpecifications({ laneWidth: 10, barrierWidth: 1, barrierChunks: 2 });
        const callback = jest.fn().mockImplementation(s => {
            expect(s).toBe(newSpecs);
        });

        const tiles = [new StraightTileModel(specs), new CurvedTileEnlargedModel(specs), new CurvedTileModel(specs)];

        editor.on('specs', callback);

        expect(editor.setSpecs(newSpecs)).toBe(editor);
        editor.load(tiles);

        expect(editor.setSpecs(newSpecs)).toBe(editor);
        expect(editor.track).toMatchSnapshot();
        expect(callback).toHaveBeenCalledTimes(2);
    });

    it('can set the rotation angle', () => {
        const editor = new TrackEditorModel(specs);

        expect(editor.track.getBuilderOption('startAngle')).toBe(0);
        expect(editor.setAngle(90)).toBe(editor);
        expect(editor.track.getBuilderOption('startAngle')).toBe(90);
    });

    it('can limit the available set of tiles', () => {
        const editor = new TrackEditorModel(specs);

        const callback = jest.fn();
        editor.on('limit', callback);

        expect(editor.limit(limitedTiles)).toBe(editor);
        expect([...editor.tiles]).toMatchSnapshot();
        expect(callback).toHaveBeenCalledTimes(1);
    });

    it('can remove the limit for the available set of tiles', () => {
        const editor = new TrackEditorModel(specs);
        editor.limit([new StraightTileModel(specs)]);

        const callback = jest.fn();
        editor.on('limit', callback);

        expect(editor.limit()).toBe(editor);
        expect([...editor.tiles]).toMatchSnapshot();
        expect(callback).toHaveBeenCalledTimes(1);
    });

    it('takes the existing track into account when limiting the available set of tiles', () => {
        const editor = new TrackEditorModel(specs);
        editor.load(track);

        const callback = jest.fn();
        editor.on('limit', callback);

        expect(editor.limit(limitedTiles)).toBe(editor);
        expect([...editor.tiles]).toMatchSnapshot();
        expect(callback).toHaveBeenCalledTimes(1);
    });

    it('can load a track', () => {
        const editor = new TrackEditorModel(specs);

        const callback = jest.fn();
        editor.on('load', callback);

        expect(editor.load(track)).toBe(editor);
        expect([...editor.track]).toMatchSnapshot();

        expect(callback).toHaveBeenCalledTimes(1);
    });

    it('updates the set of tiles when loading a track', () => {
        const editor = new TrackEditorModel(specs);
        editor.limit(limitedTiles);

        expect(editor.load(track)).toBe(editor);
        expect([...editor.tiles]).toMatchSnapshot();
    });

    it('resets the set of tiles when loading a track', () => {
        const editor = new TrackEditorModel(specs);
        editor.load(track);
        editor.limit(limitedTiles);

        expect(editor.load(track.slice(0, 1))).toBe(editor);
        expect([...editor.tiles]).toMatchSnapshot();
    });

    it('can save a track', () => {
        const editor = new TrackEditorModel(specs);
        editor.load(track);

        const callback = jest.fn().mockImplementation(data => {
            expect(data).toStrictEqual(track);
        });
        editor.on('save', callback);

        expect(editor.save()).toStrictEqual(track);

        expect(callback).toHaveBeenCalledTimes(1);
    });

    it('can clear the editor', () => {
        const editor = new TrackEditorModel(specs);
        editor.load(track);

        const callback = jest.fn();
        editor.on('clear', callback);

        expect(editor.clear()).toBe(editor);
        expect([...editor.track]).toStrictEqual([]);

        expect(callback).toHaveBeenCalledTimes(1);
    });

    it('resets the set of tiles when clearing the editor', () => {
        const editor = new TrackEditorModel(specs);
        editor.limit(limitedTiles);
        editor.load(track);

        expect(editor.clear()).toBe(editor);
        expect([...editor.track]).toStrictEqual([]);
        expect([...editor.tiles]).toMatchSnapshot();
    });

    it('updates the set of tiles when adding a tile to the track', () => {
        const editor = new TrackEditorModel(specs);
        editor.limit(limitedTiles);

        editor.track.append();
        expect([...editor.tiles]).toMatchSnapshot();
    });

    it('updates the set of tiles when removing a tile from the track', () => {
        const editor = new TrackEditorModel(specs);
        editor.limit(limitedTiles);
        editor.load(track);

        editor.track.remove(editor.track.last().id);
        expect([...editor.tiles]).toMatchSnapshot();
    });

    it('can validate an object is an instance of the class', () => {
        const counter = new TrackEditorModel(specs);
        expect(() => TrackEditorModel.validateInstance(counter)).not.toThrow();
        expect(() => TrackEditorModel.validateInstance({})).toThrow(
            'The object must be an instance of TrackEditorModel!'
        );
    });
});
