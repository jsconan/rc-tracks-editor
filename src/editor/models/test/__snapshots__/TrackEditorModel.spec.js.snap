// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackEditorModel can limit the available set of tiles 1`] = `
[
  [
    "straight-tile-1",
    2,
  ],
  [
    "curved-tile-enlarged-1",
    2,
  ],
  [
    "curved-tile-1",
    2,
  ],
]
`;

exports[`TrackEditorModel can load a track 1`] = `
[
  CurvedTileModel {
    "direction": "left",
    "id": "id-6",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
      "maxRatio": 4,
      "unlockRatio": false,
    },
  },
  StraightTileModel {
    "direction": "right",
    "id": "id-7",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
      "maxRatio": 4,
      "unlockRatio": false,
    },
  },
  CurvedTileModel {
    "direction": "right",
    "id": "id-8",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
      "maxRatio": 4,
      "unlockRatio": false,
    },
  },
]
`;

exports[`TrackEditorModel can remove the limit for the available set of tiles 1`] = `
[
  [
    "straight-tile-1",
    Infinity,
  ],
  [
    "curved-tile-1",
    Infinity,
  ],
  [
    "curved-tile-2",
    Infinity,
  ],
  [
    "curved-tile-3",
    Infinity,
  ],
  [
    "curved-tile-4",
    Infinity,
  ],
  [
    "curved-tile-enlarged-1",
    Infinity,
  ],
]
`;

exports[`TrackEditorModel can set the specifications of the tile models 1`] = `
TrackModel {
  "builder": TrackBuilder {
    "builder": [Function],
    "delegate": [Function],
    "emit": [Function],
    "listens": [Function],
    "off": [Function],
    "on": [Function],
    "once": [Function],
    "options": {
      "hPadding": 10,
      "startAngle": 0,
      "vPadding": 10,
    },
  },
  "counter": Map {
    "straight-tile-1" => 1,
    "curved-tile-enlarged-1" => 1,
    "curved-tile-1" => 1,
  },
  "counterStore": TileCounterStore {},
  "delegate": [Function],
  "emit": [Function],
  "listens": [Function],
  "modelsStore": TileModelsStore {},
  "off": [Function],
  "on": [Function],
  "once": [Function],
  "specs": TileSpecifications {
    "barrierChunks": 2,
    "barrierWidth": 1,
    "laneWidth": 10,
    "maxRatio": 4,
    "unlockRatio": false,
  },
  "tiles": TileList {
    "delegate": [Function],
    "emit": [Function],
    "list": [
      StraightTileModel {
        "direction": "right",
        "id": "id-0",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 2,
          "barrierWidth": 1,
          "laneWidth": 10,
          "maxRatio": 4,
          "unlockRatio": false,
        },
      },
      CurvedTileEnlargedModel {
        "direction": "right",
        "id": "id-1",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 2,
          "barrierWidth": 1,
          "laneWidth": 10,
          "maxRatio": 4,
          "unlockRatio": false,
        },
      },
      CurvedTileModel {
        "direction": "right",
        "id": "id-2",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 2,
          "barrierWidth": 1,
          "laneWidth": 10,
          "maxRatio": 4,
          "unlockRatio": false,
        },
      },
    ],
    "listens": [Function],
    "off": [Function],
    "on": [Function],
    "once": [Function],
    "specs": TileSpecifications {
      "barrierChunks": 2,
      "barrierWidth": 1,
      "laneWidth": 10,
      "maxRatio": 4,
      "unlockRatio": false,
    },
  },
  "tilesStore": TileListStore {},
}
`;

exports[`TrackEditorModel resets the set of tiles when clearing the editor 1`] = `
[
  [
    "straight-tile-1",
    2,
  ],
  [
    "curved-tile-enlarged-1",
    2,
  ],
  [
    "curved-tile-1",
    2,
  ],
]
`;

exports[`TrackEditorModel resets the set of tiles when loading a track 1`] = `
[
  [
    "straight-tile-1",
    2,
  ],
  [
    "curved-tile-enlarged-1",
    2,
  ],
  [
    "curved-tile-1",
    1,
  ],
]
`;

exports[`TrackEditorModel takes the existing track into account when limiting the available set of tiles 1`] = `
[
  [
    "straight-tile-1",
    1,
  ],
  [
    "curved-tile-enlarged-1",
    2,
  ],
]
`;

exports[`TrackEditorModel updates the set of tiles when adding a tile to the track 1`] = `
[
  [
    "straight-tile-1",
    1,
  ],
  [
    "curved-tile-enlarged-1",
    2,
  ],
  [
    "curved-tile-1",
    2,
  ],
]
`;

exports[`TrackEditorModel updates the set of tiles when loading a track 1`] = `
[
  [
    "straight-tile-1",
    1,
  ],
  [
    "curved-tile-enlarged-1",
    2,
  ],
]
`;

exports[`TrackEditorModel updates the set of tiles when removing a tile from the track 1`] = `
[
  [
    "straight-tile-1",
    1,
  ],
  [
    "curved-tile-enlarged-1",
    2,
  ],
  [
    "curved-tile-1",
    1,
  ],
]
`;
